proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,507 @@
+
+Each
+[FederationDomain](https://github.com/vmware-tanzu/pinniped/blob/main/generated/1.23/README.adoc#federationdomain)
+configured in the Pinniped Supervisor is an OIDC Provider issuer which implements
+the [OIDC authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth).
+
+Today, the Pinniped Supervisor only allows one hardcoded OIDC client, called `pinniped-cli`. This client is only allowed
+to redirect authcodes to the CLI's localhost listener. This makes it intentionally impossible for this client to be used
+by a webapp running on the cluster (or anywhere else on the network). The `pinniped-cli` client is implicitly available
+on all FederationDomains configured in the Supervisor, since every FederationDomain allows users to authenticate
+themselves via the Pinniped CLI for `kubectl` integration.
+
+## Terminology / Concepts
+
+- See the [definition of a "client" in the OAuth 2.0 spec](https://datatracker.ietf.org/doc/html/rfc6749#section-1.1).
+  For the purposes of this proposal, a "client" is roughly equal to a webapp which wants to know the authenticated
+  identity of a user, and may want to perform actions as that user on clusters. An admin needs to allow the client to
+  learn about the identity of the users by registering the client with the Pinniped Supervisor.
+- See also the [OIDC terminology in the OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#Terminology).
+- The OIDC clients proposed in this document are "dynamic" in the sense that they can be configured and reconfigured on
+  a running Supervisor by the admin.
+
+## Proposal
+
+### Goals and Non-goals
+
+Goals for this proposal:
+
+- Allow Pinniped admins to configure applications (OIDC clients) other than the Pinniped CLI to interact with the
+  Supervisor.
+- Provide a mechanism which governs a client's access to the token exchange APIs. Not all webapps should have permission
+  to act on behalf of the user with the Kubernetes API of the clusters, so an admin must be able to configure which
+  clients have this permission.
+- Provide a mechanism for requesting access to different aspects of a user identity, especially getting group
+  memberships or not, to allow the admin to exclude this potentially information for clients which do not need it.
+- Support a web UI based LDAP/ActiveDirectory login screen. This is needed to avoid having webapps handle the user's
+  password, which must only be seen by the Supervisor and the LDAP server. However, the details of this item have been
+  split out to a separate proposal document.
+- Client secrets must be stored encrypted or hashed, not in plain text.
+- Creation of client credentials on the operator's behalf - the server must generate any secrets.
+- The operator must be able to initiate manual rotation of client credentials.
+- Documentation describing the token exchanges a webapp backend must perform to interact with the Kubernetes API.
```

Comment:
Yes, we're considering such a feature, but it was not included in the scope of this proposal to keep the scope smaller. It might be considered as a follow-up proposal. Is that a feature that you would like to see? (cc @anjaltelang)

Reply:
Yes - I think if that feature was provided, we could completely remove our `pinniped-proxy` service (as much as I've enjoyed creating and working on a small Rust project :) )